app/javascript/mastodon/features/emoji/hooks.ts

```diff
@@ -36,25 +35,23 @@ export function useEmojify(text: string, extraEmojis?: CustomEmojiMapArg) {
   }, [extraEmojis]);
 
   const emojify = useCallback(
-    async (input: string) => {
+    (input: string) => {
       const wrapper = document.createElement('div');
       wrapper.innerHTML = input;
-      const result = await emojifyElement(wrapper, appState, extra);
-      if (result) {
-        setEmojifiedText(result.innerHTML);
-      } else {
-        setEmojifiedText(input);
-      }
+      return emojifyElement(wrapper, appState, extra);
     },
     [appState, extra],
   );
   useLayoutEffect(() => {
-    if (isModernEmojiEnabled() && !!text.trim() && stringHasAnyEmoji(text)) {
-      void emojify(text);
-    } else {
-      // If no emoji or we don't want to render, fall back.
-      setEmojifiedText(text);
+    if (isModernEmojiEnabled() && !!text.trim()) {
+      const result = emojify(text);
+      if (result) {
+        setEmojifiedText(result.innerHTML);
+        return;
+      }
     }
+    // If no emoji or we don't want to render, fall back.
+    setEmojifiedText(text);
```

Comment:
One clear optimisation path here would be running this code in the `emojifiedText` state initialiser, now that it is sync.

Currently, components with this hook will always render at least twice and `emojifiedText` will always be null in the initial render (which might also be the reason it messes with our "virtualisation").

Even with async processing we could initialise the `emojifiedText` state with simply `text` (ensuring the first render is meaningful and approximates the final text length), and if there are no emoji to parse, React won't rerender.